dprint install fails with npm with --omit=optional

It seems the package incorrectly lists some dependencies as optional that it hard depends on.

```
Error: Cannot find module '@dprint/win32-x64/package.json'
npm error Require stack:
npm error - ...\node_modules\dprint\install_api.js
npm error - ...\node_modules\dprint\install.js
npm error     at Module._resolveFilename (node:internal/modules/cjs/loader:1456:15)
npm error     at require.resolve (node:internal/modules/helpers:163:19)
npm error     at Object.runInstall (...\node_modules\dprint\install_api.js:23:52)
npm error     at Object.<anonymous> (...\node_modules\dprint\install.js:4:26)
```

All optionalDependencies does is NPM still tries to install them, but accepts install failures, so the --omit=optional should be acceptable. Yet clearly they aren't optional since lacking them breaks things.

--- a/deployment/npm/build.ts
+++ b/deployment/npm/build.ts
@@ -127,6 +127,7 @@ if (!args["publish-only"]) {
     dprintDir.join(".npmignore").writeTextSync("dprint\ndprint.exe\n");
 
     // setup each binary package
+    const executableHashes: Record<string, string> = {};
     for (const pkg of packages) {
       const pkgName = getPackageNameNoScope(pkg);
       $.logStep(`Setting up @dprint/${pkgName}...`);
@@ -140,6 +141,11 @@ if (!args["publish-only"]) {
       await $.request(zipUrl).showProgress().pipeToPath(zipPath);
       await decompress(zipPath.toString(), pkgDir.toString());
       zipPath.removeSync();
+
+      // record the executable's hash so the dprint package can verify it when
+      // downloading the binary as a fallback (e.g. for `npm install --omit=optional`)
+      const executableName = pkg.os === "win32" ? "dprint.exe" : "dprint";
+      executableHashes[pkgName] = await sha256Hex(pkgDir.join(executableName).toString());
 
       // create the package.json and readme
       pkgDir.join("README.md").writeTextSync(`# @dprint/${pkgName}\n\n${pkgName} distribution of dprint.\n`);
@@ -164,6 +170,10 @@ if (!args["publish-only"]) {
         libc: pkg.libc == null ? undefined : [pkg.libc],
       });
     }
+
+    // write the executable hashes into the dprint package so the download
+    // fallback can verify the binary it fetches from the registry
+    dprintDir.join("hashes.json").writeJsonPrettySync(executableHashes);
   }
 
   // verify that the package is created correctly
@@ -224,6 +234,14 @@ if (args.publish || args["publish-only"]) {
 
   $.logStep(`Publishing dprint...`);
   await $`cd ${dprintDir} && npm publish --provenance --access public`;
+}
+
+async function sha256Hex(filePath: string) {
+  const data = await Deno.readFile(filePath);
+  const digest = await crypto.subtle.digest("SHA-256", data);
+  return Array.from(new Uint8Array(digest))
+    .map(b => b.toString(16).padStart(2, "0"))
+    .join("");
 }
 
 function getPackageNameNoScope(name: Package) {

--- a/deployment/npm/install_api.cjs
+++ b/deployment/npm/install_api.cjs
@@ -135,12 +135,54 @@ function verifyExecutableHash(target, buffer) {
 }
 
 /**
- * Downloads a url to a buffer synchronously. Node has no synchronous https
- * client, so spawn a child node process that streams the response to stdout.
+ * Downloads a url to a buffer synchronously. Prefer curl since it transparently
+ * supports proxies (http_proxy/https_proxy/no_proxy and npm's proxy config),
+ * redirects and TLS, then fall back to node for environments without curl.
  * @param url {string}
  * @returns {Buffer}
  */
 function downloadBufferSync(url) {
+  const fromCurl = downloadBufferWithCurl(url);
+  return fromCurl != null ? fromCurl : downloadBufferWithNode(url);
+}
+
+/**
+ * @param url {string}
+ * @returns {Buffer | undefined} undefined when curl isn't available
+ */
+function downloadBufferWithCurl(url) {
+  // -f: fail on http errors, -s: silent, -S: still show errors, -L: follow redirects
+  const args = ["-fsSL"];
+  const proxy = process.env.npm_config_https_proxy
+    || process.env.npm_config_proxy
+    || process.env.HTTPS_PROXY
+    || process.env.https_proxy;
+  if (proxy) {
+    args.push("--proxy", proxy);
+  }
+  const noProxy = process.env.npm_config_noproxy || process.env.NO_PROXY || process.env.no_proxy;
+  if (noProxy) {
+    args.push("--noproxy", noProxy);
+  }
+  args.push("-o", "-", url);
+  try {
+    return require("child_process").execFileSync("curl", args, { maxBuffer: 512 * 1024 * 1024 });
+  } catch (err) {
+    if (err && err.code === "ENOENT") {
+      // curl isn't installed, fall back to node
+      return undefined;
+    }
+    throw err;
+  }
+}
+
+/**
+ * Node has no synchronous https client, so spawn a child node process that
+ * streams the response to stdout. Does not support proxies.
+ * @param url {string}
+ * @returns {Buffer}
+ */
+function downloadBufferWithNode(url) {
   const script = "const https=require('https');"
     + "function f(u){https.get(u,r=>{"
     + "const s=r.statusCode;"

--- a/deployment/npm/install_api.cjs
+++ b/deployment/npm/install_api.cjs
@@ -21,11 +21,18 @@ module.exports = {
     }
 
     const target = getTarget();
-    const sourcePackagePath = path.dirname(require.resolve("@dprint/" + target + "/package.json"));
-    const sourceExecutablePath = path.join(sourcePackagePath, dprintFileName);
+    const sourceExecutablePath = resolveSourceExecutablePath(target, dprintFileName);
 
-    if (!fs.existsSync(sourceExecutablePath)) {
-      throw new Error("Could not find executable for @dprint/" + target + " at " + sourceExecutablePath);
+    if (sourceExecutablePath == null) {
+      // the @dprint/<target> optional dependency isn't installed (for example
+      // the user ran `npm install --omit=optional`), so download the binary
+      // directly from the registry like esbuild does
+      downloadExecutable(target, dprintFileName, targetExecutablePath);
+      if (os.platform() !== "win32") {
+        // chmod +x
+        chmodX(targetExecutablePath);
+      }
+      return targetExecutablePath;
     }
 
     try {
@@ -58,6 +65,158 @@ module.exports = {
     }
   },
 };
+
+/**
+ * Resolves the path to the executable provided by the @dprint/<target>
+ * optional dependency, or undefined when that package isn't installed.
+ * @param target {string}
+ * @param dprintFileName {string}
+ * @returns {string | undefined}
+ */
+function resolveSourceExecutablePath(target, dprintFileName) {
+  let sourcePackagePath;
+  try {
+    sourcePackagePath = path.dirname(require.resolve("@dprint/" + target + "/package.json"));
+  } catch {
+    // the optional dependency wasn't installed
+    return undefined;
+  }
+  const sourceExecutablePath = path.join(sourcePackagePath, dprintFileName);
+  return fs.existsSync(sourceExecutablePath) ? sourceExecutablePath : undefined;
+}
+
+/**
+ * Downloads the binary tarball for the target from the registry and extracts
+ * the executable to the destination path.
+ * @param target {string}
+ * @param dprintFileName {string}
+ * @param destinationPath {string}
+ */
+function downloadExecutable(target, dprintFileName, destinationPath) {
+  const version = require("./package.json").version;
+  const registry = (process.env.npm_config_registry || "https://registry.npmjs.org")
+    .replace(/\/+$/, "");
+  // npm tarball urls drop the scope from the file name (e.g.
+  // https://registry.npmjs.org/@dprint/win32-x64/-/win32-x64-1.0.0.tgz)
+  const tarballUrl = registry + "/@dprint/" + target + "/-/" + target + "-" + version + ".tgz";
+  console.error("[dprint] Optional dependency @dprint/" + target + " was not installed. Downloading from " + tarballUrl);
+  const tarballBuffer = downloadBufferSync(tarballUrl);
+  // files inside an npm tarball live under the "package/" directory
+  const executableBuffer = extractFileFromTarGzip(tarballBuffer, "package/" + dprintFileName);
+  verifyExecutableHash(target, executableBuffer);
+  atomicWriteFile(destinationPath, executableBuffer);
+}
+
+/**
+ * Verifies the downloaded executable against the hash recorded at build time
+ * so that we only ever run the exact binary that was published.
+ * @param target {string}
+ * @param buffer {Buffer}
+ */
+function verifyExecutableHash(target, buffer) {
+  let hashes;
+  try {
+    hashes = require("./hashes.json");
+  } catch (err) {
+    throw new Error("Could not load hashes.json to verify the downloaded binary: " + (err && err.message || err));
+  }
+  const expected = hashes[target];
+  if (typeof expected !== "string") {
+    throw new Error("No known hash for @dprint/" + target + " to verify the download against.");
+  }
+  const actual = require("crypto").createHash("sha256").update(buffer).digest("hex");
+  if (actual !== expected) {
+    throw new Error(
+      "Integrity check failed for the downloaded @dprint/" + target + " binary.\n"
+        + "  Expected sha256: " + expected + "\n"
+        + "  Actual sha256:   " + actual,
+    );
+  }
+}
+
+/**
+ * Downloads a url to a buffer synchronously. Node has no synchronous https
+ * client, so spawn a child node process that streams the response to stdout.
+ * @param url {string}
+ * @returns {Buffer}
+ */
+function downloadBufferSync(url) {
+  const script = "const https=require('https');"
+    + "function f(u){https.get(u,r=>{"
+    + "const s=r.statusCode;"
+    + "if((s===301||s===302||s===303||s===307||s===308)&&r.headers.location){f(new URL(r.headers.location,u).toString());return;}"
+    + "if(s!==200){console.error('[dprint] Unexpected status code '+s+' downloading '+u);process.exit(1);}"
+    + "r.pipe(process.stdout);"
+    + "}).on('error',e=>{console.error(String(e&&e.message||e));process.exit(1);});}"
+    + "f(process.argv[1]);";
+  return require("child_process").execFileSync(
+    process.execPath,
+    ["-e", script, url],
+    { maxBuffer: 512 * 1024 * 1024 },
+  );
+}
+
+/**
+ * Extracts a single file from a gzipped tarball buffer.
+ * @param buffer {Buffer}
+ * @param subpath {string}
+ * @returns {Buffer}
+ */
+function extractFileFromTarGzip(buffer, subpath) {
+  const zlib = require("zlib");
+  let tar;
+  try {
+    tar = zlib.gunzipSync(buffer);
+  } catch (err) {
+    throw new Error("Invalid gzip data in downloaded tarball: " + (err && err.message || err));
+  }
+  let offset = 0;
+  while (offset < tar.length) {
+    const name = readTarString(tar, offset, 100);
+    const size = parseInt(readTarString(tar, offset + 124, 12).trim(), 8);
+    offset += 512;
+    if (!isNaN(size)) {
+      if (name === subpath) {
+        return tar.subarray(offset, offset + size);
+      }
+      // entries are padded to 512 byte boundaries
+      offset += (size + 511) & ~511;
+    }
+  }
+  throw new Error("Could not find " + JSON.stringify(subpath) + " in downloaded tarball");
+}
+
+/**
+ * @param buffer {Buffer}
+ * @param offset {number}
+ * @param length {number}
+ */
+function readTarString(buffer, offset, length) {
+  return buffer.toString("utf8", offset, offset + length).replace(/\0.*$/, "");
+}
+
+/**
+ * @param destinationPath {string}
+ * @param buffer {Buffer}
+ */
+function atomicWriteFile(destinationPath, buffer) {
+  const crypto = require("crypto");
+  const rand = crypto.randomBytes(4).toString("hex");
+  const tempFilePath = destinationPath + "." + rand;
+  fs.writeFileSync(tempFilePath, buffer);
+  try {
+    fs.renameSync(tempFilePath, destinationPath);
+  } catch (err) {
+    // will maybe throw when another process had already done this
+    // so just ignore and delete the created temporary file
+    try {
+      fs.unlinkSync(tempFilePath);
+    } catch (_err2) {
+      // ignore
+    }
+    throw err;
+  }
+}
 
 /** @filePath {string} */
 function chmodX(filePath) {

